### DISCUSSION TITLE
Is it possible to expose flux2 metrics?

### DISCUSSION THREAD

**Discussion by oscarmparedes** [fluxcd/flux2#2756] [General] [answered]
Hi all!
Im searching for documentation on how to manually expose flux2 prometheus metrics and the only way Im seeing is this guide https://fluxcd.io/docs/guides/monitoring/ . Could anyone point me in the right path? thanks a lot.

**Comment by kingdonb** [comment, marked as answer]
If you are running Flux according to the bootstrap / getting started guide, Flux exports them on port 8080 in each controller by default.

If you need help to configure a PodMonitor or ServiceMonitor, there are detailed examples in the guide that you linked!

**Reply by oscarmparedes** [reply]
In my setup I had already deployed the kube-Prometheus-stack. I noticed that in the controllers the ports were already exposing the metrics, so By just deploying the pod monitor in flux2 monitoring repo the Prometheus instance discovered the pods and everything worked perfect. 
I was confused with the guide because I thought those commands also would expose the ports, but they also install the Prometheus stack... and I had that already.

Is there any repo with alert examples other than those in the documentation?

Thanks a lot for the information and hope this helps someone else.

**Reply by kingdonb** [reply]
@oscarmparedes I understand :D Prometheus configuration is confusing and I am not happy with how mine turned out.

Here is my example where I configured AlertManager, I wanted to stay within the bounds of the Helm chart for kube-prometheus-stack and configure it through `values` exclusively if possible, ideally better would have been to configure separate resources for my own alerting and alerts configuration, but I wasn't able to divine how to do that from the Prometheus docs or kube-prometheus-stack.

So I cobbled this together from other examples I found:

https://github.com/kingdonb/flux2/blob/68624ae3d44a8513abd736ef7f592008ea4655e0/manifests/monitoring/kube-prometheus-stack/release.yaml#L25-L35

The example continues on for quite a while, some of this is boilerplate that I had to repeat in my `values` even though it was already mentioned in the defaults for the chart, because of some weird semantics around array merging that in short it's not possible to do any sort of array merging inside of a HelmRelease `spec.values` because the three-way merge doesn't provide for it. So to add my own alerts, I had to include the baseline alerts for Alert Manager. (There has to be a better way...)

I hope this configuration example helps! (I sincerely hope someone else knows better how to do this than what I figured out here, I would like to contribute further examples to the Flux monitoring guide as I think alerting is a critical function of Prometheus, and I think it is an important gap in our docs now... but Prometheus is also outside of the Flux project, so it's possible perhaps likely we consider this out of scope and leave it up to Prometheus to provide better docs.
